temporian/core/operators/join.py

```diff
@@ -0,0 +1,182 @@
+# Copyright 2021 Google LLC.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Join operator class and public API function definitions."""
+
+from typing import Optional
+
+from temporian.core.data.dtype import DType
+from temporian.core import operator_lib
+from temporian.core.data.node import (
+    Node,
+    create_node_with_new_reference,
+    Feature,
+)
+from temporian.core.operators.base import Operator
+from temporian.proto import core_pb2 as pb
+from temporian.core.data.schema import Schema
+
+JOIN_LEFT = "left"
+
+
+class Join(Operator):
+    def __init__(
+        self,
+        left: Node,
+        right: Node,
+        how: str = "left",
+        on: Optional[str] = None,
+    ):
+        super().__init__()
+
+        self.add_input("left", left)
+        self.add_input("right", right)
+        self.add_attribute("how", how)
+        if on is not None:
+            self.add_attribute("on", on)
+        self._on = on
+
+        left.schema.check_compatible_index(right.schema, "left and right")
+
+        if how not in [JOIN_LEFT]:
+            raise ValueError(
+                f"Non supported join type {how}. Supported join type(s) are:"
+                f" {JOIN_LEFT}"
+            )
+
+        if on is not None:
+            for node, node_name in [(left, "left"), (right, "right")]:
+                feature_names = node.schema.feature_names()
+                if on not in feature_names:
+                    raise ValueError(
+                        f'Feature "{on}" does not exist in {node_name}'
+                    )
+                on_dtype = node.schema.features[feature_names.index(on)].dtype
+                if on_dtype != DType.INT64:
```

Comment:
nit: add TODO for STRING join (?)

Reply:
We can effectively already do joins with the indexes.
Integer join is espetially useful when compiled with enumerate.
Lets keep an eye if we need string join.